### PR TITLE
Support for undocking and docking tabs as windows

### LINKED ISSUE #4: drag tab outside should be result into an new window :)
nice to have i think :)

link maybe help: 
https://github.com/vivlim/egui-glow-multiwin

### PR DESCRIPTION
**Pull request by Vickerinox**
Meant to close #4

## Additions:
-New widget used when hovering dragged tabs over nodes, inspired by https://github.com/Adanos020/egui_dock/issues/4#issuecomment-1595775081

-Lets the user Undock/Dock tabs as windows if a suitable placement in one of the nodes is not found, these are categorized for now as "detached branches" of the ``Tree`` type.

-new public method called ``add_window`` for ``Tree`` which adds a tab as a window.

## Todo
-Let windows contain multiple tabs and nodes instead of a singular tab

### CODE REVIEW

File: examples/simple.rs [hunk 1]
```diff
@@ -36,9 +36,10 @@ impl Default for MyApp {
         let mut tree = Tree::new(vec!["tab1".to_owned(), "tab2".to_owned()]);
 
         // You can modify the tree before constructing the dock
-        let [a, b] = tree.split_left(NodeIndex::root(), 0.3, vec!["tab3".to_owned()]);
-        let [_, _] = tree.split_below(a, 0.7, vec!["tab4".to_owned()]);
-        let [_, _] = tree.split_below(b, 0.5, vec!["tab5".to_owned()]);
+        let [a, b] =
+            tree[SurfaceIndex::root()].split_left(NodeIndex::root(), 0.3, vec!["tab3".to_owned()]);
```
Comment (Adanos020) [inline examples/simple.rs:40]: I think it would be cool if `Tree` had a `root` method!

File: src/abstract_tree/mod.rs [hunk 2]
```diff
@@ -0,0 +1,246 @@
+use egui::Rect;
+
+use crate::{
+    window::WindowState, Node, NodeIndex, NodeTree, Split, SurfaceIndex, TabDestination, TabIndex,
+};
+
+/// The basis for egui_dock
+/// This tree starts with a collection of surfaces, that then breaks down into nodes, and then into tabs.
+/// Indexing it will yield the NodeTree inside the surface which was indexed (panicing if it doesn't exist)
+pub struct Tree<Tab> {
```
Comment (Adanos020) [inline src/abstract_tree/mod.rs:10]: I don't think it makes a lot of sense to call it `Tree` anymore. It's a collection of trees with some additional state.

How about we call this `DockState` and rename `NodeTree` to just `Tree`?

Comment (Vickerinox) [reply src/abstract_tree/mod.rs]: fixed in https://github.com/Adanos020/egui_dock/pull/149/commits/59213c2104013dbf956078c04138508915e0ece3

**Comment by Adanos020** [comment]
I made a couple observations.

First, I think it would be better if the squares containing each overlay icon detected dragging, instead of only the icons.
image

It also seems a bit weird that if you drag a tab above a node behind a window, the overlay isn't fully visible. You have to move the window if you want to access all possible splits.
image

The latter seems a lot trickier to solve since we haven't yet discussed the desired behaviour in corner cases, and it would probably require a very diffferent implementation.

File: src/dock_state/tree/mod.rs [hunk 3]
```diff
@@ -180,25 +180,31 @@ impl<Tab> Tree<Tab> {
...
+    /// Returns the number of nodes in the [`Tree`].
+    /// 
+    /// This includes [`Empty`](crate::Node::Empty) nodes.
     #[inline(always)]
     pub fn len(&self) -> usize {
         self.tree.len()
     }
 
-    /// Returns `true` if the number of nodes in the tree is 0, `false` otherwise.
+    /// Returns [`true`](bool) if the number of nodes in the tree is 0, otherwise [`false`](bool).
```
Comment (Adanos020) [inline src/dock_state/tree/mod.rs:191]: These links to `bool` are unnecessary, see how Rust's core library handles bools: https://doc.rust-lang.org/nightly/src/core/option.rs.html#602

Comment (Vickerinox) [reply src/dock_state/tree/mod.rs]: ah! ill get that removed then, https://github.com/Adanos020/egui_dock/pull/149/commits/95185afafc3ea44474e2b1c4ed811bfe7b6160cb

File: src/lib.rs [hunk 4]
```diff
@@ -91,6 +72,79 @@
...
+//! A [`Surface`] is an abstraction for any tab hierarchy. There are two kinds of
+//! non-empty surfaces: `Main` and `Window`.
+//!
+//! There can only be one `Main` surface. It's the one surface that is rendered inside the
+//! [`Ui`](egui::Ui) you've passed to [`DockArea::show_inside`], or inside the
+//! [`CentralPanel`](egui::CentralPanel) created by [`DockArea::show`].
+//!
+//! On the other hand, there can be multiple `Window` surfaces. Those represent surfaces that were
+//! created by undocking tabs from the `Main` surface, and each of them is rendered inside
+//! a [`Window`](egui::Window) hence their name.
```
Comment (Vickerinox) [inline src/lib.rs:87]: perhaps clarify here explicitly that it's an ``egui::Window`` in the link, as any reader would need to either hover or click the link in order to fully understand the sentence

Comment (Adanos020) [reply src/lib.rs]: That's done now, thanks!

**Pull request merged**